### PR TITLE
feat(ingest): SageMaker feature store ingestion

### PR DESCRIPTION
**Pull request by kevinhu**
Models and jobs soon to follow!

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/examples/recipes/sagemaker_to_datahub.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+source:
+  type: sagemaker
+  config:
+    aws_region: "us-west-2"
```
Comment (hsheth2) [inline metadata-ingestion/examples/recipes/sagemaker_to_datahub.yml:4]: might want to make it clear that AWS creds are autodetected - just a comment would be enough

File: metadata-ingestion/src/datahub/ingestion/source/sagemaker.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        mce = MetadataChangeEvent(proposedSnapshot=feature_group_snapshot)
+        return MetadataWorkUnit(id=feature_group_name, mce=mce)
+
+    def get_feature_type(self, aws_type: str, feature_name: str) -> str:
+
+        mapped_type = {
+            "String": MLFeatureDataType.TEXT,
+            "Integral": MLFeatureDataType.ORDINAL,
+            "Fractional": MLFeatureDataType.CONTINUOUS,
+        }.get(aws_type)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sagemaker.py:154]: can we extract this to a module-level constant

File: metadata-ingestion/src/datahub/ingestion/source/sagemaker.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        feature_groups = self.get_all_feature_groups()
+
+        for feature_group in feature_groups:
+
+            feature_group_details = self.get_feature_group_details(
+                feature_group["FeatureGroupName"]
+            )
+
+            for feature in feature_group_details["FeatureDefinitions"]:
+                yield self.get_feature_wu(feature_group_details, feature)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sagemaker.py:286]: make sure to call `self.report.report_workunit()` as well (ditto below)

we might want to make the `report_workunit(wu)` return the workunit, so we can just do `yield self.report.report_workunit(self.get_feature_wu(...))`

File: metadata-ingestion/tests/unit/test_sagemaker_source.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+def sagemaker_source() -> SagemakerSource:
+    return SagemakerSource(
+        ctx=PipelineContext(run_id="sagemaker-source-test"),
+        config=SagemakerSourceConfig(aws_region="us-west-2"),
+    )
+
+
+@freeze_time(FROZEN_TIME)
+def test_sagemaker_ingest(tmp_path, pytestconfig):
```
Comment (hsheth2) [inline metadata-ingestion/tests/unit/test_sagemaker_source.py:27]: probably makes sense to keep this in the `tests/unit/sagemaker` directory as well

**Review by hsheth2 — CHANGES_REQUESTED** [review]
Commented on some minor things, but otherwise looks good

File: metadata-models/src/main/pegasus/com/linkedin/ml/metadata/MLFeatureTableProperties.pdl [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
+import com.linkedin.common.CustomProperties
 
 /**
  * Properties associated with a MLFeatureTable
  */
 @Aspect = {
   "name": "mlFeatureTableProperties"
 }
-record MLFeatureTableProperties {
+record MLFeatureTableProperties includes CustomProperties {
```
Comment (dexter-mh-lee) [inline metadata-models/src/main/pegasus/com/linkedin/ml/metadata/MLFeatureTableProperties.pdl:12]: You should add CustomProperties to the aspect union

File: metadata-ingestion/src/datahub/ingestion/source/aws_common.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        aws_session_token=credentials.get("SessionToken"),
+    )
+
+    assumed_role_object = sts_client.assume_role(
+        RoleArn=role_arn, RoleSessionName="DatahubIngestionSource"
+    )
+    return assumed_role_object["Credentials"]
+
+
+class AwsSourceConfig(ConfigModel):
```
Comment (dexter-mh-lee) [inline metadata-ingestion/src/datahub/ingestion/source/aws_common.py:28]: This is awesome!! 
nit: Should we have a directory for all AWS ingest that shares this config?

Comment (kevinhu) [reply metadata-ingestion/src/datahub/ingestion/source/aws_common.py:28]: Good idea – I'll add it as a comment for now

File: metadata-ingestion/src/datahub/ingestion/source/sagemaker.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+                        feature["FeatureName"],
+                    )
+                    for feature in feature_group_details["FeatureDefinitions"]
+                    if feature["FeatureName"]
+                    != feature_group_details["RecordIdentifierFeatureName"]
+                ],
+                mlPrimaryKeys=[
+                    builder.make_ml_primary_key_urn(
+                        feature_group_name,
+                        feature_group_details["RecordIdentifierFeatureName"],
```
Comment (dexter-mh-lee) [inline metadata-ingestion/src/datahub/ingestion/source/sagemaker.py:132]: Is this always one feature in SageMaker?

Comment (kevinhu) [reply metadata-ingestion/src/datahub/ingestion/source/sagemaker.py:132]: Yep – they make you specify one when a feature group is created

File: metadata-models/src/main/pegasus/com/linkedin/ml/metadata/MLFeatureTableProperties.pdl (hunk 5 above)
Comment (kevinhu) [reply metadata-models/src/main/pegasus/com/linkedin/ml/metadata/MLFeatureTableProperties.pdl:12]: Punting this for now until we fix CustomProperties in general

**Review by hsheth2 — APPROVED** [review]
LGTM

**Review by shirshanka — APPROVED** [review]
LGTM

**Pull request merged**